packages/ui/react-ui-editor/src/components/TextEditor/TextEditor.tsx

```diff
@@ -39,93 +40,143 @@ export type TextEditorRef = {
   view?: EditorView;
 };
 
+export type TextEditorSlots = {
+  root?: Omit<ComponentProps<'div'>, 'ref'>;
+  editor?: {
+    className?: string;
+    placeholder?: string;
+    spellCheck?: boolean;
+    tabIndex?: number;
+    theme?: ThemeStyles;
+  };
+};
+
 export type TextEditorProps = {
-  model?: EditorModel;
+  model: EditorModel;
   extensions?: Extension[];
-  theme?: {
-    [selector: string]: StyleSpec;
-  };
-  slots?: EditorSlots;
-  onKeyDown?: (event: KeyboardEvent, info: CursorInfo) => void;
-} & Pick<HTMLAttributes<HTMLDivElement>, 'onBlur' | 'onFocus'>;
+  slots?: TextEditorSlots;
+  editorMode?: EditorMode;
+};
 
 /**
- * Simple text editor.
+ * Base text editor.
+ * NOTE: Rather than adding properties, try to create extensions that can be reused.
  */
-export const TextEditor = forwardRef<TextEditorRef, TextEditorProps>(
-  ({ model, extensions = [], theme = defaultStyles, slots = {}, onKeyDown, ...props }, forwardedRef) => {
-    const { id, content } = model ?? {};
+export const BaseTextEditor = forwardRef<TextEditorRef, TextEditorProps>(
+  ({ model, extensions = [], slots, editorMode }, forwardedRef) => {
     const { themeMode } = useThemeContext();
+    const tabsterDOMAttribute = useFocusableGroup({ tabBehavior: 'limited' });
 
     const [parent, setParent] = useState<HTMLDivElement | null>(null);
     const [state, setState] = useState<EditorState>();
     const [view, setView] = useState<EditorView>();
-
-    // TODO(burdon): The ref may be instantiated before the view is created.
     useImperativeHandle(
       forwardedRef,
-      () => {
-        return {
-          editor: parent,
-          state,
-          view,
-        };
-      },
-      [view],
+      () => ({
+        editor: parent,
+        state,
+        view,
+      }),
+      [view, state, parent],
     );
 
+    // TODO(burdon): Pass in extension?
+    const collaboration = useCollaboration(model, themeMode);
+
     useEffect(() => {
       if (!parent) {
         return;
       }
 
-      view?.destroy();
-
       const state = EditorState.create({
-        doc: content?.toString(),
+        doc: model.content?.toString(),
         extensions: [
-          bracketMatching(),
-          closeBrackets(),
-          placeholder(slots.editor?.placeholder ?? ''),
-          EditorView.lineWrapping,
-
-          // Themes.
-          EditorView.theme(theme),
-          ...(themeMode === 'dark'
-            ? [syntaxHighlighting(oneDarkHighlightStyle)]
-            : [syntaxHighlighting(defaultHighlightStyle)]),
+          // TODO(burdon): Factor out VIM mode?
```

Comment:
if we can find a way to make the additional focus-management that's added in this component extensible then this should be possible...would certainly clean this base component up